kafka/src/main/java/org/apache/calcite/adapter/kafka/KafkaStreamTable.java

```diff
@@ -77,14 +87,61 @@
         if (tableOptions.getConsumerParams() != null) {
           consumerConfig.putAll(tableOptions.getConsumerParams());
         }
-        Consumer consumer = new KafkaConsumer<>(consumerConfig);
-        consumer.subscribe(Collections.singletonList(tableOptions.getTopicName()));
-
+        Consumer consumer;
+        if (tableOptions.getConsumer() != null) {
+          consumer = tableOptions.getConsumer();
+        } else {
+          consumer = new KafkaConsumer<>(consumerConfig);
+        }
+        if (tableOptions.getTimestamp() != -1) {
+          // seek to special timestamp
+          consumer.subscribe(Collections.singletonList(tableOptions.getTopicName()),
+              new ConsumerRebalanceListener() {
+                @Override public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+                }
+
+                @Override public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+                  // consumer can seek to special offset after the consumer join the consumer group
+                  seekOnTimestamp(consumer, tableOptions.getTopicName(),
+                      tableOptions.getTimestamp());
+                }
+              });
+        } else {
+          consumer.subscribe(Collections.singletonList(tableOptions.getTopicName()));
+        }
         return new KafkaMessageEnumerator(consumer, tableOptions.getRowConverter(), cancelFlag);
       }
     };
   }
 
+  /**
+   * seek to special timestamp
+   *
```

Comment:
We should start the comment with uppercase letter, also the comment should end up with dot.

Reply:
I have modified these comment.